config/auth/iap/service_account_auth_helper.go

```diff
@@ -0,0 +1,140 @@
+
+package config
+
+import (
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/json"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"golang.org/x/oauth2/jws"
+
+	oauth "golang.org/x/oauth2/google"
+)
+
+const (
+	audURL                  = "https://www.googleapis.com/oauth2/v4/token"
+	jwtBearerTokenGrantType = "urn:ietf:params:oauth:grant-type:jwt-bearer"
+	expirationSecs          = 3600
+)
+
+// GetIDTokenWithServiceAccount gets an IAP ID Token required for authenticating a service account with IAP.
+// For more info, see https://cloud.google.com/iap/docs/authentication-howto#authenticating_from_a_service_account
+func GetIDTokenWithServiceAccount(config IapConfig) (string, error) {
+	creds, err := ioutil.ReadFile(config.ServiceAccountKeyPath)
+	if err != nil {
+		return "", fmt.Errorf("could not read service account creds file: %v", err)
+	}
+
+	// Creating and signing JWT token.
+
+	jwtc, err := oauth.JWTConfigFromJSON(creds)
+	if err != nil {
+		return "", fmt.Errorf("could not parse config from service account creds: %v", err)
+	}
+
+	pk, err := parseKey(jwtc.PrivateKey)
+	if err != nil {
+		return "", fmt.Errorf("invalid private json key: %v", err)
+	}
+
+	now := time.Now().Unix()
+
+	claims := &jws.ClaimSet{
+		Aud: audURL,
+		Iss: jwtc.Email,
+		Sub: jwtc.Email,
+		Iat: now,
+		Exp: now + expirationSecs,
+		PrivateClaims: map[string]interface{}{
+			"target_audience": config.IapClientId,
+		},
+	}
+
+	h := &jws.Header{
+		Algorithm: "RS256",
+		Typ:       "JWT",
+		KeyID:     jwtc.PrivateKeyID,
+	}
+
+	jwt, err := jws.Encode(h, claims, pk)
+	if err != nil {
+		return "", fmt.Errorf("failed to encode jwt: %v", err)
+	}
+
+	// Use the JWT to get the OIDC token.
+
+	form := url.Values{}
+	form.Add("grant_type", jwtBearerTokenGrantType)
+	form.Add("assertion", jwt)
+
+	req, err := http.NewRequest("POST", audURL, strings.NewReader(form.Encode()))
+	if err != nil {
+		return "", fmt.Errorf("failed to create request: %v", err)
+	}
+	req.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+
+	client := &http.Client{}
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return "", fmt.Errorf("failed to get OIDC token: %s", err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return "", fmt.Errorf("failed to get successful response: %+v", resp)
+	}
+
+	tokenBody, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return "", fmt.Errorf("could not read response body of OIDC token request")
+	}
+
+	var payload map[string]interface{}
+	err = json.Unmarshal(tokenBody, &payload)
+	if err != nil {
+		return "", fmt.Errorf("failed to parse access token payload %q: %v", string(tokenBody), err)
+	}
+
+	return fmt.Sprintf("%v", payload["id_token"]), nil
+}
+
+func parseKey(key []byte) (*rsa.PrivateKey, error) {
+	block, _ := pem.Decode(key)
+	if block != nil {
+		key = block.Bytes
+	}
+	parsedKey, err := x509.ParsePKCS8PrivateKey(key)
+	if err != nil {
```

Comment:
This winds up blackholing the initial `err`, which I'm not a fan of. Is there any way to inspect the key to figure out which PKCS scheme to use?

Reply:
Doesn't seem like there's an easy way in the x509 package besides directly trying to parse the key :(